reframe/core/schedulers/lsf.py

```diff
@@ -30,45 +32,57 @@ def __init__(self):
             f'schedulers/@{self.registered_name}/job_submit_timeout'
         )
 
+    def _format_option(self, var, option):
+        if var is not None:
+            return self._prefix + ' ' + option.format(var)
+        else:
+            return ''
+
     def emit_preamble(self, job):
-        num_tasks_per_node = job.num_tasks_per_node or 1
-        num_nodes = job.num_tasks // num_tasks_per_node
 
         preamble = [
-            self._format_option(f'-J {job.name}'),
-            self._format_option(f'-o {job.stdout}'),
-            self._format_option(f'-e {job.stderr}'),
-            self._format_option(f'-nnodes {num_nodes}')
+            self._format_option(job.name, '-J "{0}"'),
+            self._format_option(job.stdout, '-o {0}'),
+            self._format_option(job.stderr, '-e {0}'),
+            self._format_option(job.num_tasks, '-n {0}'),
+            self._format_option(
+                min(job.num_tasks * job.num_cpus_per_task,
+                    job.num_tasks_per_node * job.num_cpus_per_task),
+                '-R "span[ptile={0}]"'
+            ),
         ]
 
         # add job time limit in minutes
         if job.time_limit is not None:
             preamble.append(
-                self._format_option(f'-W {int(job.time_limit // 60)}')
+                f'{self._prefix} -W {int(job.time_limit // 60)}'
             )
 
+        for opt in job.sched_access:
+            preamble.append('%s %s' % (self._prefix, opt))
+
         # emit the rest of the options
         options = job.options + job.cli_options
         for opt in options:
             if opt.startswith('#'):
                 preamble.append(opt)
             else:
-                preamble.append(self._format_option(opt))
+                preamble.append(self._prefix + ' ' + opt)
 
-        # change to working dir with cd
-        preamble.append(f'cd {job.workdir}')
+        if job.sched_exclusive_access:
+            preamble.append(f'{self._prefix} -x')
 
         return preamble
 
     def submit(self, job):
-        cmd = f'bsub {job.script_filename}'
-        completed = _run_strict(cmd, timeout=self._submit_timeout)
+        with open(job.script_filename, 'r') as f:
+            completed = subprocess.run(
+                args='bsub', stdin=f, capture_output=True)
         jobid_match = re.search(r'^Job <(?P<jobid>\S+)> is submitted',
-                                completed.stdout)
+                                completed.stdout.decode('utf-8'))
```

Comment:
Why do you need to decode the string here?

Reply:
Without calling decode my jobs crash immediately with the following error:

> Traceback (most recent call last):
  File "/lustre/scafellpike/local/HT03340/hxt01/mxt16-hxt01/reframe/reframe/frontend/executors/__init__.py", line 267, in _safe_call
    return fn(*args, **kwargs)
  File "/lustre/scafellpike/local/HT03340/hxt01/mxt16-hxt01/reframe/reframe/core/hooks.py", line 80, in _fn
    func(obj, *args, **kwargs)
  File "/lustre/scafellpike/local/HT03340/hxt01/mxt16-hxt01/reframe/reframe/core/pipeline.py", line 2292, in run
    super().run()
  File "/lustre/scafellpike/local/HT03340/hxt01/mxt16-hxt01/reframe/reframe/core/pipeline.py", line 1680, in run
    self._job.submit()
  File "/lustre/scafellpike/local/HT03340/hxt01/mxt16-hxt01/reframe/reframe/core/schedulers/__init__.py", line 425, in submit
    return self.scheduler.submit(self)
  File "/lustre/scafellpike/local/HT03340/hxt01/mxt16-hxt01/reframe/reframe/core/schedulers/lsf.py", line 81, in submit
    jobid_match = re.search(r'^Job <(?P<jobid>\S+)> is submitted',
  File "/lustre/scafellpike/local/apps/python3/3.8.0/lib/python3.8/re.py", line 199, in search
    return _compile(pattern, flags).search(string)
TypeError: cannot use a string pattern on a bytes-like object